docs/adr/0008-use-tracing.md

```diff
@@ -0,0 +1,165 @@
+# Replace logging and error reporting infrastructure with tracing.
+
+* Status: proposed
+* Deciders:
+  * AppServices team: ?
+  * Firefox Android: ?
+  * Firefox iOS: ?
+  * Firefox Desktop: ?
+
+* Date: Mar 06, 2024
+* Feedback deadline: Mar 20, 2024
+
+## Context and Problem Statement
+
+Currently all Rust code in [app-services](https://github.com/mozilla/application-services/)
+uses the `log` crate for debugging and diagnostics.
+Log output is often captured during tests where it's just printed to stdout/stderr if the test
+fails. Further, both our mobile apps and Firefox Desktop have techniques for having this log
+output be sent across the FFI so it can be captured in whatever "native" logging system is
+in place (eg, logcat on Android, custom file logging on iOS, Log.sys.mjs/Console.sys.mjs logging on Desktop)
+
+application-services also has an [error support crate](https://github.com/mozilla/application-services/tree/main/components/support/error),
+designed explicitly for error reporting for the applications.
+Android and iOS both hook into this to report errors to Sentry, whereas Desktop has not yet implemented this functionality.
+While the error support crate has capabilities beyond just moving the error across the FFI, this capability is the focus of this document,
+and any changes proposed to this crate by this document are limited to just this.
+
+### Problems with the current approaches.
+
+#### Problems using `log`
+
+The main problem with the `log` module is a design choice made by that crate,
+which is the concept of a [global "max level"](https://docs.rs/cli-log/latest/cli_log/fn.set_max_level.html)
+
+In Gecko, simply setting this max level to, say, `Trace` can cause performance regressions in other
+crates - eg, [see this desktop bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1874215) for one example.
```

Comment:
>  Use log::debug! if you want runtime configuration for the logging. Use release_log::debug! if you want to ensure maximum performance.

How does that work for crates we don't control? If those 2 conditions were reversed I could see it, but I don't see how that would be viable today.

Reply:
Hmmm, I'm also wondering how this will work! `release_log` would need to clone the `log` crate's interface, right—it couldn't just use the `log` crate as-is?

Reply:
> How does that work for crates we don't control? If those 2 conditions were reversed I could see it, but I don't see how that would be viable today.

It would mean we need to go through the m-c repo and change `log::debug` to `release_log::debug` (and also trace).  That's definitely a strike against it, but it seems doable.

For 3rd-party crates, there's nothing we could do.  But I would hope that a library crate wouldn't make expensive logging calls with the assumption that consumers would compile them out.

> Hmmm, I'm also wondering how this will work! release_log would need to clone the log crate's interface, right—it couldn't just use the log crate as-is?

I was thinking that we could forward the calls.  `release_log::info!` would either be a no-op or call `log::info!`

Reply:
TBH, I really don't think a solution which involves all crates *other* than ours changing their logging strategy for our benefit is viable enough to include here as an option. Let me know if you feel strongly that I should though.

Reply:
> I was thinking that we could forward the calls. `release_log::info!` would either be a no-op or call `log::info!`

Ahhh, gotcha—I missed that this would involve having all Gecko crates adopt `release_log`, not just ours. That does feel pretty challenging to do, even if we don't think about third-party crates.

Reply:
Thanks - I'm happy to change that 2nd para, although I think I should leave off that last sentence:

> Changing all the existing code is a difficult task -- especially since much of that code is vendored in from other Mozilla repositories.

I don't think "difficult task" goes far enough - if we want that sentence at all, I think it would read something like "Changing all the existing code and creating new expectations for all future Rust code in m-c so that application-services logging works would be unwelcome and we could expect siginificant pushback, if not an outright veto"

Reply:
> The way I see it, the logging design is not a deal-breaker by itself. It's flawed, but there are ways to work around it.

I think we need to keep in mind that from the POV of all those existing crates, it's not flawed - it works perfectly fine and as intended.

> Maybe that would be a good time for them to start using some sort of debug-only logging macro.

Why would they want to do that? That sounds like a solution to a problem they don't have. In that desktop bug there was no suggestion that the http3 crate was doing the wrong, or even sub-optimal thing, which I agree with.

Reply:
> Why would they want to do that? That sounds like a solution to a problem they don't have. In that desktop bug there was no suggestion that the http3 crate was doing the wrong, or even sub-optimal thing, which I agree with.

If http3 wanted to move over to `tracing` they would need to do something since there's no longer a max log level.  Checking if they're in a release build seems simple to implement.  It compiles out the entire check, which might make a difference (for example, I think checking the level for the `http3` target would require taking a lock on the hash map).  The downside is that you can't enable those logs in a release build with an ENV variable, but that's how it currently works and they're okay with that.

Reply:
Yeah - that's exactly what I mean. There's no reason to believe they want to move to tracing and they are OK with how it currently works. That's why I don't see any reason for this doc to suggest that now might be a good time for them (or indeed, anyone else) to change anything.

Reply:
Agreed.  I'm not exactly sure how to word it, but what I like about the proposal is that there's no pressure for them to change anything and also would be easy for them to migrate to the new system if they wanted to.